File: packages/core-ts/src/core/normalize.ts
/**
 * Message normalization for PrivacyLens.
 *
 * Converts any supported message format into a canonical list of message dicts.
 */

/** A canonical message dict with at least role and content. */
export interface Message {
  role: string;
  content: string;
  [key: string]: unknown;
}

/** Anthropic-style input: a dict with a "messages" key. */
interface AnthropicStyleInput {
  messages: Message[];
  [key: string]: unknown;
}

/** All supported input types for normalizeMessages. */
export type MessageInput = string | Message[] | AnthropicStyleInput;

/**
 * Normalize any supported message format to a canonical list of message dicts.
 *
 * Supported input formats:
 * - Plain string → `[{ role: "user", content: input }]`
 * - OpenAI-style array of message dicts → returned unchanged
 * - Anthropic-style object with a `messages` key → `input.messages`
 *
 * @param input - The message input in any supported format.
 * @returns A list of message dicts, each containing at least `role` and `content`.
 * @throws {TypeError} If the input is not a string, array, or recognized object structure.
 */
export function normalizeMessages(input: MessageInput): Message[] {
  if (typeof input === "string") {
    return [{ role: "user", content: input }];
  }

  if (Array.isArray(input)) {
    return input;
  }

  if (typeof input === "object" && input !== null) {
    if ("messages" in input) {
      return (input as AnthropicStyleInput).messages;
    }
    throw new TypeError(
      `Unrecognized object format: expected an object with a 'messages' key ` +
        `(Anthropic-style), but got keys: ${JSON.stringify(Object.keys(input))}`
    );
  }

  throw new TypeError(
    `Unsupported message format: expected a string, array, or object with a 'messages' key, ` +
      `but got ${typeof input}`
  );
}

File: packages/core-py/src/privacylens/core/normalize.py
"""Message normalization for PrivacyLens.

Converts any supported message format into a canonical list of message dicts.
"""

from __future__ import annotations

from typing import Any


def normalize_messages(input: Any) -> list[dict[str, Any]]:
    """Normalize any supported message format to a canonical list of message dicts.

    Supported input formats:
    - Plain string → ``[{"role": "user", "content": input}]``
    - OpenAI-style list of dicts (each with ``role`` and ``content``) → returned unchanged
    - Anthropic-style dict with a ``"messages"`` key → ``input["messages"]``

    Args:
        input: The message input in any supported format.

    Returns:
        A list of message dicts, each containing at least ``role`` and ``content`` keys.

    Raises:
        TypeError: If the input is not a string, list, or recognized dict structure.
    """
    if isinstance(input, str):
        return [{"role": "user", "content": input}]

    if isinstance(input, list):
        return input

    if isinstance(input, dict):
        if "messages" in input:
            return input["messages"]
        raise TypeError(
            f"Unrecognized dict format: expected a dict with a 'messages' key "
            f"(Anthropic-style), but got keys: {list(input.keys())!r}"
        )

    raise TypeError(
        f"Unsupported message format: expected a str, list, or dict with a 'messages' key, "
        f"but got {type(input).__name__!r}"
    )
